### PR TITLE
feat(tokens): Add token UPDATE route

### PR DESCRIPTION
**Pull request by ian-fox**
* `POST` requests to `/tokens/{id}` will update the specified token
* Name can't be changed such that it conflicts with another token

Related to https://github.com/screwdriver-cd/screwdriver/issues/532

### CODE REVIEW

File: test/plugins/tokens.test.js [hunk 1]
```diff
@@ -16,7 +16,7 @@ const getTokenMock = (token) => {
 
     mock.update = sinon.stub();
     mock.regenerate = sinon.stub();
-    mock.toJson = sinon.stub().returns(token);
+    mock.toJson = sinon.stub().returns(hoek.clone(token));
```
Comment (ian-fox) [inline test/plugins/tokens.test.js]: The fact that the `GET` endpoint calls `toJson` and then `delete`s some of the fields was causing problems with other tests. It should have been like this all along.

File: plugins/tokens/update.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                    // console.log(user.id);
+                    throw boom.forbidden('User does not own token');
+                }
+
+                return user.tokens
+                    .then((tokens) => {
+                        // Make sure it won't cause a name conflict
+                        const match = tokens && tokens.find(t => t.name === request.payload.name);
+
+                        if (match && request.payload.name && request.params.id !== match.id) {
```
Comment (FenrirUnbound) [inline plugins/tokens/update.js]: Why do you need to check for `request.payload.name`?

Comment (ian-fox) [inline plugins/tokens/update.js]: Oh actually that's a good point. I was thinking if they didn't include a new name there couldn't be a name conflict, but in that case there wouldn't be a `match` in the first place.

File: plugins/tokens/update.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                if (!token) {
+                    throw boom.notFound('Token does not exist');
+                }
+
+                if (!user) {
+                    throw boom.notFound('User does not exist');
+                }
+
+                if (token.userId !== user.id) {
+                    // console.log(token.userId);
```
Comment (tkyi) [inline plugins/tokens/update.js]: `console.log`

File: plugins/tokens/update.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
+'use strict';
+
+const boom = require('boom');
+const joi = require('joi');
+const schema = require('screwdriver-data-schema');
+const idSchema = joi.reach(schema.models.token.base, 'id');
+
+module.exports = () => ({
+    method: 'PUT',
```
Comment (FenrirUnbound) [inline plugins/tokens/update.js]: `PATCH` might be more appropriate

Comment (ian-fox) [inline plugins/tokens/update.js]: Today I learned that there are http methods other than `GET`, `POST`, `PUT`, and `DELETE`, and that `PUT` requires the full resource be in the request.

File: test/plugins/tokens.test.js [hunk 5]
```diff
@@ -185,6 +189,8 @@ describe('token plugin test', () => {
                     scope: ['user']
                 }
             };
+
+            userMock.tokens = tokensGetterMock([tokenMock]);
```
Comment (FenrirUnbound) [inline test/plugins/tokens.test.js]: Not needed. This is redundant of https://github.com/screwdriver-cd/screwdriver/pull/573/files#diff-ee5dae69c9de402833a459d19c780d18R67

File: test/plugins/tokens.test.js [hunk 6]
```diff
@@ -218,24 +225,119 @@ describe('token plugin test', () => {
...
                     scope: ['user']
+                },
+                payload: {
+                    name: testToken.name,
+                    description: 'a new description'
                 }
             };
 
             tokenFactoryMock.get.resolves(tokenMock);
+            userMock.tokens = tokensGetterMock([tokenMock]);
```
Comment (FenrirUnbound) [inline test/plugins/tokens.test.js]: Not needed. This is redundant of https://github.com/screwdriver-cd/screwdriver/pull/573/files#diff-ee5dae69c9de402833a459d19c780d18R67

File: test/plugins/tokens.test.js [hunk 7]
```diff
@@ -218,24 +225,119 @@ describe('token plugin test', () => {
...
+            userFactoryMock.get.withArgs({ username }).resolves(userMock);
```
Comment (FenrirUnbound) [inline test/plugins/tokens.test.js]: Redundant of https://github.com/screwdriver-cd/screwdriver/pull/573/files#diff-ee5dae69c9de402833a459d19c780d18R74

File: test/plugins/tokens.test.js [hunk 8]
```diff
@@ -218,24 +225,119 @@ describe('token plugin test', () => {
...
+            tokenMock.toJson.returns(expected);
+
+            return server.inject(options).then((reply) => {
+                assert.equal(reply.statusCode, 200);
+                assert.calledOnce(tokenMock.update);
+                assert.deepEqual(reply.result, expected);
+            });
+        });
+
+        it('doesn\'t error if the new name conflicts only with the old name', () => {
```
Comment (FenrirUnbound) [inline test/plugins/tokens.test.js]: From a testing perspective, this looks no different than this test https://github.com/screwdriver-cd/screwdriver/pull/573/files#diff-ee5dae69c9de402833a459d19c780d18R276

File: test/plugins/tokens.test.js [hunk 9]
```diff
@@ -218,24 +225,119 @@ describe('token plugin test', () => {
...
                     username,
                     scope: ['user']
+                },
+                payload: {
+                    name: testToken.name,
+                    description: 'a new description'
                 }
             };
 
             tokenFactoryMock.get.resolves(tokenMock);
```
Comment (FenrirUnbound) [inline test/plugins/tokens.test.js]: This line is also redundant

File: test/plugins/tokens.test.js [hunk 10]
```diff
@@ -218,24 +225,119 @@ describe('token plugin test', () => {
...
+                url: `/tokens/${tokenId}`,
+                credentials: {
+                    username,
+                    scope: ['user']
+                }
+            };
+
+            tokenFactoryMock.get.resolves(tokenMock);
+            userMock.tokens = tokensGetterMock([tokenMock]);
+            userFactoryMock.get.withArgs({ username }).resolves(userMock);
```
Comment (FenrirUnbound) [inline test/plugins/tokens.test.js]: All these are defined at the top-level `beforeEach` so there isn't a need to re-defining the same behavior here

File: test/plugins/tokens.test.js (hunk 1 above)
Comment (FenrirUnbound) [inline test/plugins/tokens.test.js]: When did we introduce `hoek`? For cloning and combining object maps, ES6 has [`Object.assign`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign)

File: plugins/tokens/update.js (hunk 4 above)
Comment (FenrirUnbound) [inline plugins/tokens/update.js]: It's a rabbit-hole of fun 😸

File: test/plugins/tokens.test.js (hunk 1 above)
Comment (ian-fox) [inline test/plugins/tokens.test.js]: Ooh, did not know that existed. Will switch to that now.

File: test/plugins/tokens.test.js (hunk 10 above)
Comment (ian-fox) [inline test/plugins/tokens.test.js]: Ah, there's where my confusion was. I thought that beforeEach ran before each of the `describe` blocks. Didn't realize it also ran before `it` blocks, so I thought I needed to have that there to restore functionality after doing things like having the `userFactoryMock` resolve `null` to test the 404 case.

File: plugins/tokens/update.js [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+                if (!user) {
+                    throw boom.notFound('User does not exist');
+                }
+
+                if (token.userId !== user.id) {
+                    throw boom.forbidden('User does not own token');
+                }
+
+                return user.tokens
+                    .then((tokens) => {
```
Comment (FenrirUnbound) [inline plugins/tokens/update.js]: Let's keep the indentation consistent. Some of the `.then` statements are indented and some aren't.

Comment (ian-fox) [inline plugins/tokens/update.js]: Whoops, missed the one without an indent.

**Pull request merged**